src/lib/core/datetime/date-adapter.ts

```diff
@@ -0,0 +1,196 @@
+   *     greater than 11, it should roll into the previous / next year.
+   * @param date The date of month of the date. If `date` is less than 1 or greater than the number
+   *     of days in the `month`, it should roll into the previous / next month.
+   * @returns The new date.
+   */
+  abstract createDate(year: number, month: number, date: number): D;
+
+  /**
+   * Gets today's date.
+   * @returns Today's date.
+   */
+  abstract today(): D;
+
+  /**
+   * Parses a date from a value.
+   * @param value The value to parse.
+   * @param fmt The expected format of the value being parsed (type is implementation-dependent).
+   * @returns The parsed date, or null if date could not be parsed.
+   */
+  abstract parse(value: any, fmt?: any): D | null;
+
+  /**
+   * Formats a date as a string.
+   * @param date The value to parse.
+   * @param fmt The format to use for the result string.
+   * @returns The parsed date, or null if date could not be parsed.
+   */
+  abstract format(date: D, fmt?: any): string;
+
+  /**
+   * Adds the given number of years to the date. Years are counted as if flipping 12 pages on the
+   * calendar for each year and then finding the closest date in the new month. For example when
+   * adding 1 year to Feb 29, 2016, the resulting date will be Feb 28, 2017.
+   * @param date The date to add years to.
+   * @param years The number of years to add (may be negative).
+   * @returns A new date equal to the given one with the specified number of years added.
+   */
+  abstract addCalendarYears(date: D, years: number): D;
+
+  /**
+   * Adds the given number of months to the date. Months are counted as if flipping a page on the
+   * calendar for each month and then finding the closest date in the new month. For example when
+   * adding 1 month to Jan 31, 2017, the resulting date will be Feb 28, 2017.
+   * @param date The date to add months to.
+   * @param months The number of months to add (may be negative).
+   * @returns A new date equal to the given one with the specified number of months added.
+   */
+  abstract addCalendarMonths(date: D, months: number): D;
+
+  /**
+   * Adds the given number of days to the date. Days are counted as if moving one cell on the
+   * calendar for each day.
+   * @param date The date to add days to.
+   * @param days The number of days to add (may be negative).
+   * @returns A new date equal to the given one with the specified number of days added.
+   */
+  abstract addCalendarDays(date: D, days: number): D;
+
+  /**
+   * Sets the locale used for all dates.
+   * @param locale The new locale.
+   */
+  setLocale(locale: any) {
+    this.locale = locale;
+  }
+
+  /**
+   * Clones the given date.
+   * @param date The date to clone
+   * @returns A new date equal to the given date.
+   */
+  clone(date: D): D {
+    return this.createDate(this.getYear(date), this.getMonth(date), this.getDate(date));
```

Comment:
Would it be better to leave this without an implementation here so that the method can be used for all of date/time/datetime?